src/plugins/ui_actions/public/actions/dynamic_action_manager.ts

```diff
@@ -0,0 +1,304 @@
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { v4 as uuidv4 } from 'uuid';
+import { Subscription } from 'rxjs';
+import { ActionStorage, SerializedEvent } from './dynamic_action_storage';
+import { UiActionsService } from '../service';
+import { SerializedAction } from './types';
+import { TriggerContextMapping } from '../types';
+import { ActionDefinition } from './action';
+import { defaultState, transitions, selectors, State } from './dynamic_action_manager_state';
+import { StateContainer, createStateContainer } from '../../../kibana_utils';
+
+const compareEvents = (
+  a: ReadonlyArray<{ eventId: string }>,
+  b: ReadonlyArray<{ eventId: string }>
+) => {
+  if (a.length !== b.length) return false;
+  for (let i = 0; i < a.length; i++) if (a[i].eventId !== b[i].eventId) return false;
+  return true;
+};
+
+export type DynamicActionManagerState = State;
+
+export interface DynamicActionManagerParams {
+  storage: ActionStorage;
+  uiActions: Pick<
+    UiActionsService,
+    'registerAction' | 'attachAction' | 'unregisterAction' | 'detachAction' | 'getActionFactory'
+  >;
+  isCompatible: <C = unknown>(context: C) => Promise<boolean>;
+}
+
+export class DynamicActionManager {
+  static idPrefixCounter = 0;
+
+  private readonly idPrefix = `D_ACTION_${DynamicActionManager.idPrefixCounter++}_`;
+  private stopped: boolean = false;
+  private reloadSubscription?: Subscription;
+
+  /**
+   * UI State of the dynamic action manager.
+   */
+  protected readonly ui = createStateContainer(defaultState, transitions, selectors);
+
+  constructor(protected readonly params: DynamicActionManagerParams) {}
+
+  protected getEvent(eventId: string): SerializedEvent {
+    const oldEvent = this.ui.selectors.getEvent(eventId);
+    if (!oldEvent) throw new Error(`Could not find event [eventId = ${eventId}].`);
+    return oldEvent;
+  }
+
+  /**
+   * We prefix action IDs with a unique `.idPrefix`, so we can render the
+   * same dashboard twice on the screen.
+   */
+  protected generateActionId(eventId: string): string {
+    return this.idPrefix + eventId;
+  }
+
+  protected reviveAction(event: SerializedEvent) {
+    const { eventId, triggers, action } = event;
+    const { uiActions, isCompatible } = this.params;
+    const { name } = action;
+
+    const actionId = this.generateActionId(eventId);
+    const factory = uiActions.getActionFactory(event.action.factoryId);
+    const actionDefinition: ActionDefinition<any> = {
+      ...factory.create(action as SerializedAction<object>),
```

Comment:
I've moved `getDisplayName` to factory. `isCompatible` needs to be controlled by the manager so it makes sure action is displayed in the right places. `id`, in general, I believe can be set by action factories, but in drilldowns it is overwritten here because drilldown manager needs to know precisely the IDs and make sure they don't collide, to be able to register and unregister actions as the dynamic action manager is started and stopped.

Reply:
This is also why we need `registerDrilldown`, because users of Drilldowns should not need to learn those internal details.